src/gstring.rs

```diff
@@ -0,0 +1,333 @@
+    }
+
+    pub fn from(s: &[u8]) -> Self {
+        let cstr = unsafe { CString::from_vec_unchecked(s.to_vec()) };
+        Self::from_cstring(cstr)
+    }
+
+    pub fn from_cstring(s: CString) -> Self {
+        Self { s: Some(s), ptr: ptr::null(), length: 0 }
+    }
+
+    pub fn from_ptr_container(ptr: *const *const c_char) -> Vec<Self> {
+        unsafe {
+            let num = c_ptr_array_len(ptr);
+            if num == 0 || ptr.is_null() {
+                return Vec::new();
+            }
+
+            let mut res = Vec::with_capacity(num);
+            for i in 0..num {
+                let len = libc::strlen(*ptr.add(i));
+                let s = std::slice::from_raw_parts(*ptr.add(i) as *const u8, len + 1);
+                res.push(GString::from(s));
+            }
+            res
+        }
+    }
+
+    pub fn as_str(&self) -> &str {
+        let cstr = match self.s {
+            Some(ref s) => s.as_c_str(),
+            None => unsafe {
+                assert!(!self.ptr.is_null());
+                assert!(self.length > 0);
+                let bytes = std::slice::from_raw_parts(self.ptr as *const u8, self.length + 1);
+                CStr::from_bytes_with_nul_unchecked(bytes)
+            }
+        };
+        cstr.to_str().unwrap()
+    }
+
+    pub fn to_cstring(&self) -> CString {
+        if let Some(ref s) = self.s {
+            let bytes = s.as_bytes_with_nul();
+            unsafe { CString::from_vec_unchecked(bytes.to_vec()) }
+        } else {
+            assert!(!self.ptr.is_null());
+            unsafe { CString::from_raw(self.ptr as *mut i8) }
+        }
+    }
+}
+
+impl Drop for GString {
+    fn drop(&mut self) {
+        if self.s.is_none() {
+            unsafe {
+                glib_ffi::g_free(self.ptr as *mut _);
+            }
+        }
+    }
+}
+
+impl fmt::Debug for GString {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "[GString for {}]", self.as_str())
+    }
+}
+
+impl fmt::Display for GString {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}", self.as_str())
+    }
+}
+
+impl std::borrow::Borrow<str> for GString {
+    fn borrow(&self) -> &str {
+        self.as_str()
+    }
+}
+
+impl Ord for GString {
+    fn cmp(&self, other: &GString) -> Ordering {
+        self.as_str().bytes().cmp(other.as_str().bytes())
+    }
+}
+
+impl PartialOrd for GString {
+    fn partial_cmp(&self, other: &GString) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl PartialEq for GString {
+    fn eq(&self, other: &GString) -> bool {
+        self.as_str() == other.as_str()
+    }
+}
+
+impl PartialEq<GString> for std::string::String {
+    fn eq(&self, other: &GString) -> bool {
+        self.as_str() == other.as_str()
+    }
+}
+
+impl PartialEq<str> for GString {
+    fn eq(&self, other: &str) -> bool {
+        self.as_str() == other
+    }
+}
+
+impl PartialEq<&'static str> for GString {
```

Comment:
This one seems unneeded, but it would probably make sense to also implement `PartialOrd` for `str` and `String` (and implement them the other way around too)

Reply:
It's still needed for code like:
```
error[E0277]: can't compare `glib::GString` with `&str`                                                                                                                                                            
   --> gstreamer-pbutils/src/encoding_profile.rs:521:9                                                                                                                                                             
    |                                                                                                                                                                                                              
521 |         assert_eq!(audio_profile.get_name().unwrap(), AUDIO_PROFILE_NAME);                                                                                                                                   
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ no implementation for `glib::GString == &str`                                                                                     
    |                                                                                                                                                                                                              
    = help: the trait `std::cmp::PartialEq<&str>` is not implemented for `glib::GString`                                                                             
```

Reply:
Strange, I would've expected the impl just above to cover this case. But in any case, instead of static this should be possible with any lifetime. `impl<'a> PartialEq<&'a str> ....`